### PR TITLE
Do not enumerate unavailable digests and ciphers in list -*-commands

### LINKED ISSUE #13594: list -cipher-commands lists ciphers that aren't available
The list option `-cipher-commands` option is supposed to list the available ciphers. However it takes no note of what providers are available so, for example, if the  legacy provider is not loaded it will still list legacy ciphers.

### PR DESCRIPTION
**Pull request by beldmit**
Fixes #13594 

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: apps/list.c [hunk 1]
```diff
@@ -958,6 +1016,16 @@ static void list_type(FUNC_TYPE ft, int one)
     for (fp = functions; fp->name != NULL; fp++) {
         if (fp->type != ft)
             continue;
+	switch (ft) {
+	case FT_cipher:
+	    if (is_cipher_available(fp->name) == 0) continue;
```
[the remark this answers is not in the document]
Comment (beldmit) [reply apps/list.c]: Done.

File: apps/list.c [hunk 2]
```diff
@@ -1295,8 +1363,10 @@ const OPTIONS list_options[] = {
     {"select", OPT_SELECT_NAME, 's', "Select a single algorithm"},
     {"commands", OPT_COMMANDS, '-', "List of standard commands"},
     {"standard-commands", OPT_COMMANDS, '-', "List of standard commands"},
+#ifndef OPENSSL_NO_DEPRECATED_3_0
     {"digest-commands", OPT_DIGEST_COMMANDS, '-',
```
[the remark this answers is not in the document]
Comment (beldmit) [reply apps/list.c:1347]: I've updated the man page and CHANGES.md. I can add the warning message either, but it's absent for `-engine` option.

File: apps/list.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (beldmit) [reply apps/list.c]: Thanks!

File: apps/list.c [hunk 3]
```diff
@@ -945,6 +945,64 @@ static void list_options_for_command(const char *command)
...
+    ERR_set_mark();
+    md = EVP_MD_fetch(NULL, name, NULL);
+    ERR_pop_to_mark();
+    if (md != NULL) {
+        EVP_MD_free(md);
+        return 1;
+    }
+
+#ifndef OPENSSL_NO_DEPRECATED_3_0
+#ifndef OPENSSL_NO_ENGINE
```
[the remark this answers is not in the document]
Comment (beldmit) [reply apps/list.c]: Done. Thanks!

**Comment by beldmit** [comment]
CI minimal failure is relevant, but I don't know how to properly deal with it. The failing test takes the output of `-cipher-commands` and plans tests according to it. 

It can either be fixed either via updating via apps/progs.pl or hard-coding.

File: apps/list.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattcaswell) [reply apps/list.c:1347]: IMO if people aren't told that something is deprecated when they attempt to use it, then its not really deprecated.

File: apps/lib/engine.c [hunk 4]
```diff
@@ -163,3 +163,29 @@ int get_legacy_pkey_id(OSSL_LIB_CTX *libctx, const char *algname, ENGINE *e)
 
     return pkey_id;
 }
+
+const EVP_MD *get_digest_from_engine(const char *name)
+{
+#ifndef OPENSSL_NO_ENGINE
+    ENGINE *eng;
```
Comment (mattcaswell) [inline apps/lib/engine.c:170]: Blank line after declarations

File: apps/lib/engine.c [hunk 5]
```diff
@@ -163,3 +163,29 @@ int get_legacy_pkey_id(OSSL_LIB_CTX *libctx, const char *algname, ENGINE *e)
...
+        return EVP_get_digestbyname(name);
+    }
+#endif
+    return NULL;
+}
+
+const EVP_CIPHER *get_cipher_from_engine(const char *name)
+{
+#ifndef OPENSSL_NO_ENGINE
+    ENGINE *eng;
```
Comment (mattcaswell) [inline apps/lib/engine.c:183]: Blank line

File: apps/list.c [hunk 6]
```diff
@@ -945,6 +945,44 @@ static void list_options_for_command(const char *command)
...
+/* Look through providers' digests*/
+    ERR_set_mark();
+    md = EVP_MD_fetch(NULL, name, NULL);
+    ERR_pop_to_mark();
+    if (md != NULL) {
+        EVP_MD_free(md);
+        return 1;
+    }
+
+#ifndef OPENSSL_NO_DEPRECATED_3_0
```
Comment (mattcaswell) [inline apps/list.c:961]: I don't think this guard is necessary. Won't `get_digest_from_engine()` just return NULL in a no-deprecated build?

File: doc/man1/openssl-list.pod.in [hunk 7]
```diff
@@ -71,13 +73,17 @@ Display a list of standard commands.
 
 =item B<-digest-commands>
 
+This option is deprecated.
```
Comment (mattcaswell) [inline doc/man1/openssl-list.pod.in:76]: Please state what users should use instead

File: doc/man1/openssl-list.pod.in [hunk 8]
```diff
@@ -71,13 +73,17 @@ Display a list of standard commands.
...
+
 Display a list of message digest commands, which are typically used
 as input to the L<openssl-dgst(1)> or L<openssl-speed(1)> commands.
 
 =item B<-cipher-commands>
 
+This option is deprecated.
```
Comment (mattcaswell) [inline doc/man1/openssl-list.pod.in:83]: and here

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
> The failing test takes the output of -cipher-commands and plans tests according to it.

Can we rewrite the test to not use `-cipher-commands`, e.g. use `-cipher-algorithms` instead? We shouldn't be using deprecated stuff in our tests.

File: apps/list.c (hunk 6 above)
Comment (beldmit) [reply apps/list.c:961]: I can move this #ifdef to the function, if it makes the code clearer

**Comment by beldmit** [comment]
> Can we rewrite the test to not use `-cipher-commands`, e.g. use `-cipher-algorithms` instead? We shouldn't be using deprecated stuff in our tests.

The `-cipher-algorithms` is not equivalent option. In fact, there is no strictly equivalent option.

May be we should not deprecate these options and just add the checks for presence of the algorithms?

File: apps/list.c (hunk 6 above)
Comment (mattcaswell) [reply apps/list.c:961]: I think it can be removed entirely. If you're in a no-deprecated build, then you have no-engine implicitly.

**Comment by mattcaswell** [comment]
> May be we should not deprecate these options and just add the checks for presence of the algorithms?

Hmm. I'm not keen on that. We really don't want to have this hard coded list for any longer than we have to.

**Comment by beldmit** [comment]
Ok. I have some ideas how to eliminate the command from tests and hopefully will be able to do it

**Comment by paulidale** [comment]
My suggestion would be to condition that test out in a deprecated build.  I don't think it is a relevant test anymore -- it was created as a sanity check.

File: apps/lib/engine.c (hunk 4 above)
Comment (beldmit) [reply apps/lib/engine.c:170]: Fixed. Thanks!

File: apps/lib/engine.c (hunk 5 above)
Comment (beldmit) [reply apps/lib/engine.c:184]: Fixed. Thanks!

File: apps/list.c (hunk 6 above)
Comment (beldmit) [reply apps/list.c]: Fixed. Thanks!

File: doc/man1/openssl-list.pod.in (hunk 8 above)
Comment (beldmit) [reply doc/man1/openssl-list.pod.in]: Fixed. Thanks!

File: doc/man1/openssl-list.pod.in (hunk 7 above)
Comment (beldmit) [reply doc/man1/openssl-list.pod.in]: Fixed. Thanks!

File: apps/list.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (beldmit) [reply apps/list.c:1341]: Done. Thanks!

File: apps/list.c [hunk 9]
```diff
@@ -945,6 +945,38 @@ static void list_options_for_command(const char *command)
     BIO_printf(bio_out, "- -\n");
 }
 
+static int is_md_available(const char *name)
+{
+    EVP_MD *md;
+
+/* Look through providers' digests */
```
Comment (paulidale) [inline apps/list.c:952]: The comment should be indented.

File: apps/list.c [hunk 10]
```diff
@@ -945,6 +945,38 @@ static void list_options_for_command(const char *command)
...
+    }
+
+    return (get_digest_from_engine(name) == NULL) ? 0 : 1;
+}
+
+static int is_cipher_available(const char *name)
+{
+    EVP_CIPHER *cipher;
+
+/* Look through providers' ciphers */
```
Comment (paulidale) [inline apps/list.c:968]: Likewise, indent.

File: apps/list.c [hunk 11]
```diff
@@ -958,6 +990,16 @@ static void list_type(FUNC_TYPE ft, int one)
     for (fp = functions; fp->name != NULL; fp++) {
         if (fp->type != ft)
             continue;
+	switch (ft) {
```
Comment (paulidale) [inline apps/list.c:993]: The indentation for this block looks off.

File: apps/list.c [hunk 12]
```diff
@@ -958,6 +990,16 @@ static void list_type(FUNC_TYPE ft, int one)
...
+	case FT_cipher:
+	    if (!is_cipher_available(fp->name)) continue;
```
Comment (paulidale) [inline apps/list.c:995]: We break the continue onto the following line.

File: apps/list.c [hunk 13]
```diff
@@ -958,6 +990,16 @@ static void list_type(FUNC_TYPE ft, int one)
...
+	    break;
+	case FT_md:
+	    if (!is_md_available(fp->name)) continue;
```
Comment (paulidale) [inline apps/list.c:998]: likewise, line break required.

File: doc/man1/openssl-list.pod.in [hunk 14]
```diff
@@ -14,12 +14,14 @@ B<openssl list>
 [B<-1>]
 [B<-commands>]
 [B<-digest-commands>]
-[B<-digest-algorithms>]
+{- output_off() if $disabled{"deprecated-3.0"}; ""
+-}[B<-digest-algorithms>]
```
Comment (paulidale) [inline doc/man1/openssl-list.pod.in:18]: Where is the matching `output_on()`?

File: doc/man1/openssl-list.pod.in [hunk 15]
```diff
@@ -14,12 +14,14 @@ B<openssl list>
...
 [B<-kdf-algorithms>]
 [B<-mac-algorithms>]
 [B<-random-generators>]
 [B<-cipher-commands>]
-[B<-cipher-algorithms>]
+{- output_off() if $disabled{"deprecated-3.0"}; ""
```
Comment (paulidale) [inline doc/man1/openssl-list.pod.in:23]: Again, no `output_on()`

**Review by paulidale — CHANGES_REQUESTED** [review]
Some style nits mostly.  I'm concerned about the man page not turning output back on though.

File: apps/list.c (hunk 9 above)
Comment (beldmit) [reply apps/list.c]: Done. Thanks!

File: apps/list.c (hunk 11 above)
Comment (beldmit) [reply apps/list.c]: Done. Thanks!

File: apps/list.c (hunk 10 above)
Comment (beldmit) [reply apps/list.c]: Done. Thanks!

File: apps/list.c (hunk 12 above)
Comment (beldmit) [reply apps/list.c]: Done. Thanks!

File: apps/list.c (hunk 13 above)
Comment (beldmit) [reply apps/list.c]: Done. Thanks!

File: doc/man1/openssl-list.pod.in (hunk 14 above)
Comment (beldmit) [reply doc/man1/openssl-list.pod.in:18]: Done. Thanks!

File: doc/man1/openssl-list.pod.in (hunk 15 above)
Comment (beldmit) [reply doc/man1/openssl-list.pod.in:24]: Done. Thanks!

**Review by paulidale — APPROVED** [review]

**Comment by beldmit** [comment]
Merged. Thanks!

**Comment by romen** [comment]
@beldmit I noticed that the following now fails:

~~~
./config no-deprecated && make -s && make cmd-nits
Configuring OpenSSL version 3.0.0-alpha10-dev for target linux-x86_64
Using os-specific seed configuration
Creating configdata.pm
Running configdata.pm
Creating Makefile

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL.md file first)      ***
***                                                                ***
**********************************************************************
/usr/bin/perl ./util/find-doc-nits -c
doc/man1/openssl-list.pod: list does not implement -cipher-commands
doc/man1/openssl-list.pod: list does not implement -digest-commands
make: *** [Makefile:3652: cmd-nits] Error 1
~~~

Is this expected?

**Comment by beldmit** [comment]
No, this is not expected. Will take a look.

**Comment by beldmit** [comment]
Fix in #13728

**Comment by romen** [comment]
> No, this is not expected. Will take a look.

> Fix in #13728

Great, thanks for the quick follow-up!

**Pull request closed without merging**